src/lib/ppx_version/versioned_module.ml

```diff
@@ -195,14 +206,126 @@ let erase_stable_versions =
+  object
+    inherit Ast_traverse.map as super
+
+    method! core_type typ =
+      match typ.ptyp_desc with
+      | Ptyp_constr
+          ({ txt = Ldot (Ldot (Ldot (lid, "Stable"), vn), "t"); loc }, typs)
+        when try
+               validate_module_version vn loc ;
+               true
+             with _ -> false ->
+          (* Change [.Stable.Vn.t] to [.Stable.Vn.With_all_version_tags.t] *)
+          let typ =
+            { typ with
+              ptyp_desc =
+                Ptyp_constr
+                  ( { txt =
+                        Ldot
+                          ( Ldot
+                              ( Ldot (Ldot (lid, "Stable"), vn)
+                              , with_all_version_tags_module )
+                          , "t" )
+                    ; loc
+                    }
+                  , typs )
+            }
+          in
+          super#core_type typ
+      | _ ->
+          super#core_type typ
+
+    method! type_declaration type_decl =
+      if String.equal type_decl.ptype_name.txt "t" then
+        let loc = type_decl.ptype_name.loc in
+        { (super#type_declaration type_decl) with
+          ptype_name = { txt = "typ"; loc }
+        ; ptype_manifest =
+            (* type typ = t = ... *)
+            ( match type_decl.ptype_manifest with
+            | Some
+                ( { ptyp_desc =
+                      Ptyp_constr
+                        ( { txt = Ldot (Ldot (Ldot (lid, "Stable"), vn), "t")
+                          ; loc
+                          }
+                        , params )
+                  ; _
+                  } as manifest_ty ) ->
+                (* type typ = Foo.Stable.Vn.t becomes type typ = Foo.Stable.Vn.With_all_version_tags.t *)
+                Some
+                  { manifest_ty with
+                    ptyp_desc =
+                      Ptyp_constr
+                        ( { txt =
+                              Ldot
+                                ( Ldot
+                                    ( Ldot (Ldot (lid, "Stable"), vn)
+                                    , with_all_version_tags_module )
+                                , "t" )
+                          ; loc
+                          }
+                        , params )
+                  }
+            | Some m ->
+                Some m
+            | None ->
+                let params =
+                  List.map type_decl.ptype_params ~f:(fun (ty, _invariance) ->
+                      ty )
+                in
+                Some
+                  { ptyp_desc = Ptyp_constr ({ txt = Lident "t"; loc }, params)
+                  ; ptyp_loc = loc
+                  ; ptyp_loc_stack = []
+                  ; ptyp_attributes = []
+                  } )
+        ; ptype_attributes =
+            (let (module Ast_builder) = Ast_builder.make loc in
+             let open Ast_builder in
+             [ create_attr ~loc (Located.mk "deriving") (PStr [ [%stri bin_io] ])
```

Comment:
Doesn't this erase all of the existing attributes? It would be better not to do that, if possible.

Reply:
It does. My thought was that the module `With_all_version_tags` (and `With_top_version_tag`) should have the `Bin_prot.Binable.S` interface, and nothing more.

I had started out preserving the other derivers, but then thought it wasn't valuable. It increases compilation time, and you can use the derived code from the `Vn` module, if needed, because `Vn.With.... t = Vn.t`.